#Practice Programming Assignment : 3

###Nishita Namdeo
###200416
------------------------------------

###**Problem Statement :**
Solve the multiclass image classification problem using the CIFAR-10 dataset. 
Perform the following analysis:

1. Build a CNN to solve the problem (typical Deep Learning approach that uses a feed-forward MLP at the end of the CNN layers to classify the images).

2. Extract convolutional features and train a conventional machine learning classifier (e.g. SVM or Logistic Regression).
You need to first get the training and test data ready for this experiment. 

3. Compare the performances of the models built in step 1 and 2.


####Import Necessary Libraries

In [15]:
import tensorflow as tf
from tensorflow import keras
from sklearn.svm import SVC
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from tensorflow.keras.datasets import cifar10
import seaborn as sns
import matplotlib.pyplot as plt

####Load CIFAR-10 dataset

In [2]:
(x_train, y_train), (x_test, y_test) = cifar10.load_data()

170498071/170498071 [==============================] - 2s 0us/step


####Normalize the Data

In [3]:
x_train = x_train / 255.0
x_test = x_test / 255.0

####Define CNN Model

In [4]:
model = keras.Sequential([
    
keras.layers.Conv2D(32,(3,3), activation = "relu", input_shape = (32,32,3)),
keras.layers.MaxPooling2D((2,2)),
keras.layers.Flatten(),
keras.layers.Dense(128, activation='relu'),
keras.layers.Dense(64, activation='relu'),
keras.layers.Dense(10, activation = 'softmax')
])

####Compiling the Model

In [5]:
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])

####Train the CNN model

In [6]:
model.fit(x_train, y_train, epochs=13, validation_data = (x_test, y_test))

Epoch 1/13
1563/1563 [==============================] - 34s 21ms/step - loss: 1.4360 - accuracy: 0.4838 - val_loss: 1.2161 - val_accuracy: 0.5644
Epoch 2/13
1563/1563 [==============================] - 32s 21ms/step - loss: 1.1250 - accuracy: 0.6061 - val_loss: 1.1006 - val_accuracy: 0.6110
Epoch 3/13
1563/1563 [==============================] - 33s 21ms/step - loss: 0.9777 - accuracy: 0.6551 - val_loss: 1.0809 - val_accuracy: 0.6203
Epoch 4/13
1563/1563 [==============================] - 33s 21ms/step - loss: 0.8709 - accuracy: 0.6943 - val_loss: 1.0295 - val_accuracy: 0.6419
Epoch 5/13
1563/1563 [==============================] - 34s 21ms/step - loss: 0.7747 - accuracy: 0.7302 - val_loss: 1.0510 - val_accuracy: 0.6459
Epoch 6/13
1563/1563 [==============================] - 32s 20ms/step - loss: 0.6991 - accuracy: 0.7537 - val_loss: 1.0401 - val_accuracy: 0.6545
Epoch 7/13
1563/1563 [==============================] - 33s 21ms/step - loss: 0.6194 - accuracy: 0.7804 - val_loss: 1.0972 -

####Evaluate the CNN model on test data

In [7]:
test_loss , test_accuracy = model.evaluate(x_test,y_test)

313/313 [==============================] - 3s 8ms/step - loss: 1.5923 - accuracy: 0.6380


####Extract features from the last convolutional layer of the CNN model

In [8]:
model_features = keras.Model(inputs=model.inputs, outputs=model.layers[-2].output)
train_features = model_features.predict(x_train)
test_features = model_features.predict(x_test)

313/313 [==============================] - 2s 6ms/step


####Train and Evaluate SVM classifier on extracted features

In [9]:
svm = SVC(kernel='linear')
svm.fit(train_features, y_train)
y_pred_svm = svm.predict(test_features)
test_accuracy_svm = accuracy_score(y_test, y_pred_svm)

/usr/local/lib/python3.9/dist-packages/sklearn/utils/validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


####Train and Evaluate Logistic Regression classifier on extracted features

In [10]:
lr = LogisticRegression()
lr.fit(train_features, y_train)
y_pred_lr = lr.predict(test_features)
test_accuracy_lr = accuracy_score(y_test, y_pred_lr)

/usr/local/lib/python3.9/dist-packages/sklearn/utils/validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.9/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


####The Results

In [11]:
print('CNN Model Test Accuracy:', test_accuracy)
print('SVM Classifier Test Accuracy:', test_accuracy_svm)
print('Logistic Regression ClassifiervTest Accuracy:', test_accuracy_lr)

CNN Model Test Accuracy: 0.6380000114440918
SVM Classifier Test Accuracy: 0.6404
Logistic Regression ClassifiervTest Accuracy: 0.6432


#### - The CNN model was trained for 13 epochs, and the test accuracy was about 64%.

####- Finally, we took the features from the final convolutional layer of the CNN    model and used them to train SVM and Logistic Regression classifiers. 

####- The test accuracy for the SVM classifier was about 64%, whereas the test accuracy for the Logistic Regression classifier was around 64.3%.

####- These findings lead us to the conclusion that the Logistic Regression Model outperformed CNN and SVM models trained on the identical data.